### PR TITLE
Upgradeable batch proof method id lists in light client circuit

### LINKED ISSUE #1577: Upgradeable batch proof method id lists in light client circuit
https://github.com/chainwayxyz/citrea/issues/1566#issuecomment-2527984758

> batch proof method id —> will open issue about how this is going to be updated

- [ ] We define a new type of LightClientProofNamespace transaction, that tells the light client circuit that there is now a new batch proof method id that can be used. 
- [ ] We modify light client vars so that the batch proof method ids are stored in an array of (height, method id), just like global FORKS vector.
- [ ] In the light client circuit, we check for the new type of txs defined in step 1, if they come from a new pubkey constant (can call this METHOD_ID_UPGRADE_AUTHORITY) we update the light client state defined in step 2.

### PR DESCRIPTION
**Pull request by kpp**
# Description

## Linked Issues
- Fixes https://github.com/chainwayxyz/citrea/issues/1577

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/bitcoin-da/src/helpers/mod.rs [hunk 1]
```diff
@@ -26,6 +28,7 @@ impl TransactionKindLightClient {
             TransactionKindLightClient::Complete => 0u16.to_le_bytes().to_vec(),
             TransactionKindLightClient::Chunked => 1u16.to_le_bytes().to_vec(),
             TransactionKindLightClient::ChunkedPart => 2u16.to_le_bytes().to_vec(),
+            TransactionKindLightClient::BatchProverMethodId => 3u16.to_le_bytes().to_vec(),
```
Comment (eyusufatik) [inline crates/bitcoin-da/src/helpers/mod.rs]: rename to batchproofmethodid

File: crates/bitcoin-da/src/verifier.rs [hunk 2]
```diff
@@ -17,6 +17,9 @@ use crate::spec::BitcoinSpec;
 
 pub const WITNESS_COMMITMENT_PREFIX: &[u8] = &[0x6a, 0x24, 0xaa, 0x21, 0xa9, 0xed];
 
+pub const METHOD_ID_UPGRADE_AUTHORITY: &[u8] =
```
Comment (eyusufatik) [inline crates/bitcoin-da/src/verifier.rs]: make this a constant in the necessary circuits

File: crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs [hunk 3]
```diff
@@ -22,6 +22,15 @@ pub struct SequencerCommitment {
     pub l2_end_block_number: u64,
 }
 
+/// A new batch proof method_id starting to be applied from the l2_block_number (inclusive).
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, BorshDeserialize, BorshSerialize)]
+pub struct BatchProofMethodId {
+    /// Merkle root of soft confirmation hashes
+    pub method_id: [u32; 8],
+    /// Start L2 block's number
+    pub l2_block_number: u64,
```
Comment (eyusufatik) [inline crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs]: rename to activation_l2_height

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 4]
```diff
@@ -433,3 +439,247 @@ async fn test_light_client_proving_multiple_proofs() -> Result<()> {
...
+            .client
+            .http_client()
+            .get_light_client_proof_by_l1_height(method_id_l1_height + 1)
+            .await?;
+        let lcp_output = lcp.unwrap().light_client_proof_output;
+        assert_eq!(
+            lcp_output.batch_proof_method_ids,
+            vec![
+                (0, citrea_risc0_batch_proof::BATCH_PROOF_BITCOIN_ID),
+                (100, citrea_risc0_batch_proof::BATCH_PROOF_BITCOIN_ID)
```
Comment (eyusufatik) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: maybe updating with a different method id is smarter?

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs (hunk 3 above)
Comment (eyusufatik) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs]: also comments

**Review by eyusufatik — APPROVED** [review]
important update done so late. inshallah we don't miss anything :)

File: crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs [hunk 5]
```diff
@@ -39,11 +48,13 @@ impl core::cmp::Ord for SequencerCommitment {
 /// Data written to DA can only be one of these two types
```
Comment (kpp) [inline crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs]: `TODO: rename to da service request smth smth` can be removed

Comment (yaziciahmet) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/da.rs]: https://github.com/chainwayxyz/citrea/pull/1630/commits/267b23b939ff0b6ed426ed0aad118dd081eae5a0

**Pull request merged**